File: parser.swift
// parser.swift
import Foundation

struct RateResponse: Codable {
    let success: Bool
    let rates: [String: Double]
}

struct HistoryEntry: Codable {
    let timestamp: String
    let rates: [String: Double]
}

class RateParser {
    private let apiURL = "https://api.exchangerate.host/latest?base=USD"
    private let cacheTTL = 60
    private var rates: [String: Double] = [:]
    private var lastUpdate: Date?
    private var history: [HistoryEntry] = []

    init() {
        fetchRates()
    }

    @discardableResult
    func fetchRates() -> Bool {
        guard let url = URL(string: apiURL) else { return false }
        let semaphore = DispatchSemaphore(value: 0)
        var success = false

        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            defer { semaphore.signal() }
            guard let data = data, error == nil else {
                print("Error fetching rates: \(error?.localizedDescription ?? "unknown")")
                return
            }
            do {
                let decoder = JSONDecoder()
                let result = try decoder.decode(RateResponse.self, from: data)
                if result.success {
                    self.rates = result.rates
                    self.lastUpdate = Date()
                    self.history.append(HistoryEntry(
                        timestamp: self.lastUpdate!.ISO8601Format(),
                        rates: result.rates
                    ))
                    if self.history.count > 10 {
                        self.history.removeFirst()
                    }
                    success = true
                }
            } catch {
                print("Error parsing JSON: \(error)")
            }
        }
        task.resume()
        semaphore.wait()
        return success
    }

    func getRates() -> [String: Double] {
        if let lastUpdate = lastUpdate {
            if Date().timeIntervalSince(lastUpdate) > TimeInterval(cacheTTL) {
                fetchRates()
            }
        }
        return rates
    }

    func filterRates(currencies: [String]) -> [String: Double] {
        let allRates = getRates()
        var result: [String: Double] = [:]
        for c in currencies {
            if let rate = allRates[c] {
                result[c] = rate
            }
        }
        return result
    }

    func searchCurrency(query: String) -> [String: Double] {
        let allRates = getRates()
        var result: [String: Double] = [:]
        let q = query.lowercased()
        for (k, v) in allRates {
            if k.lowercased().contains(q) {
                result[k] = v
            }
        }
        return result
    }

    @discardableResult
    func refresh() -> Bool {
        return fetchRates()
    }

    func exportJSON(filename: String = "rates.json") -> Bool {
        let data: [String: Any] = [
            "timestamp": lastUpdate?.ISO8601Format() ?? "unknown",
            "rates": getRates()
        ]
        do {
            let json = try JSONSerialization.data(withJSONObject: data, options: .prettyPrinted)
            try json.write(to: URL(fileURLWithPath: filename))
            return true
        } catch {
            print("Export failed: \(error)")
            return false
        }
    }

    func exportCSV(filename: String = "rates.csv") -> Bool {
        let rates = getRates()
        var lines = ["Currency,Rate"]
        for (k, v) in rates.sorted(by: { $0.key < $1.key }) {
            lines.append("\(k),\(String(format: "%.4f", v))")
        }
        do {
            try lines.joined(separator: "\n").write(toFile: filename, atomically: true, encoding: .utf8)
            return true
        } catch {
            print("Export failed: \(error)")
            return false
        }
    }

    func showHistory() {
        if history.isEmpty {
            print("No history yet.")
            return
        }
        for (i, entry) in history.enumerated() {
            print("[\(i+1)] \(entry.timestamp) – \(entry.rates.count) currencies")
        }
    }

    func displayRates(rates: [String: Double]? = nil) {
        let displayRates = rates ?? getRates()
        if displayRates.isEmpty {
            print("No rates available.")
            return
        }
        let dateStr = lastUpdate?.formatted() ?? "unknown"
        print("\nRates (USD base) – updated: \(dateStr)")
        print(String(repeating: "-", count: 40))
        var count = 0
        for (k, v) in displayRates.sorted(by: { $0.key < $1.key }) {
            if count >= 20 { break }
            print("\(k.padding(toLength: 5, withPad: " ", startingAt: 0)) : \(String(format: "%.4f", v))")
            count += 1
        }
        if displayRates.count > 20 {
            print("... and \(displayRates.count - 20) more")
        }
    }
}

func main() {
    let parser = RateParser()
    print("=== Exchange Rate Parser ===")
    while true {
        print("\n1. Show all rates")
        print("2. Filter by currency")
        print("3. Search currency")
        print("4. Export to JSON")
        print("5. Export to CSV")
        print("6. Refresh rates")
        print("7. Show history")
        print("8. Exit")
        print("Choose: ", terminator: "")
        guard let choice = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) else { continue }
        switch choice {
        case "1":
            parser.displayRates()
        case "2":
            print("Enter currency codes (comma-separated): ", terminator: "")
            guard let input = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) else { break }
            let currencies = input.uppercased().split(separator: ",").map { String($0).trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
            if !currencies.isEmpty {
                let filtered = parser.filterRates(currencies: currencies)
                parser.displayRates(rates: filtered)
            } else {
                print("No currencies specified.")
            }
        case "3":
            print("Enter currency code or name: ", terminator: "")
            guard let query = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines), !query.isEmpty else {
                print("Query cannot be empty.")
                break
            }
            let found = parser.searchCurrency(query: query)
            parser.displayRates(rates: found)
        case "4":
            print("Filename (default: rates.json): ", terminator: "")
            var fname = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "rates.json"
            if fname.isEmpty { fname = "rates.json" }
            if parser.exportJSON(filename: fname) {
                print("Exported to \(fname)")
            }
        case "5":
            print("Filename (default: rates.csv): ", terminator: "")
            var fname = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "rates.csv"
            if fname.isEmpty { fname = "rates.csv" }
            if parser.exportCSV(filename: fname) {
                print("Exported to \(fname)")
            }
        case "6":
            if parser.refresh() {
                print("Rates refreshed.")
            } else {
                print("Refresh failed.")
            }
        case "7":
            parser.showHistory()
        case "8":
            print("Goodbye!")
            return
        default:
            print("Invalid choice.")
        }
    }
}

main()

File: Parser.cs
// Parser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

class RateResponse
{
    public bool Success { get; set; }
    public Dictionary<string, double> Rates { get; set; }
}

class HistoryEntry
{
    public string Timestamp { get; set; }
    public Dictionary<string, double> Rates { get; set; }
}

class RateParser
{
    private readonly string apiUrl = "https://api.exchangerate.host/latest?base=USD";
    private readonly int cacheTTL;
    private Dictionary<string, double> rates;
    private DateTime? lastUpdate;
    private List<HistoryEntry> history;
    private readonly HttpClient client;

    public RateParser(int cacheTTL = 60)
    {
        this.cacheTTL = cacheTTL;
        this.rates = new Dictionary<string, double>();
        this.history = new List<HistoryEntry>();
        this.client = new HttpClient();
        this.client.Timeout = TimeSpan.FromSeconds(5);
        _ = FetchRates();
    }

    private async Task<bool> FetchRates()
    {
        try
        {
            var response = await client.GetStringAsync(apiUrl);
            var data = JsonSerializer.Deserialize<RateResponse>(response);
            if (data != null && data.Success)
            {
                rates = data.Rates;
                lastUpdate = DateTime.Now;
                history.Add(new HistoryEntry
                {
                    Timestamp = lastUpdate.Value.ToString("o"),
                    Rates = new Dictionary<string, double>(rates)
                });
                if (history.Count > 10) history.RemoveAt(0);
                return true;
            }
            return false;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error fetching rates: {e.Message}");
            return false;
        }
    }

    public Dictionary<string, double> GetRates()
    {
        if (lastUpdate.HasValue && (DateTime.Now - lastUpdate.Value).TotalSeconds > cacheTTL)
        {
            _ = FetchRates();
        }
        return rates;
    }

    public Dictionary<string, double> FilterRates(List<string> currencies)
    {
        var allRates = GetRates();
        var result = new Dictionary<string, double>();
        foreach (var c in currencies)
        {
            if (allRates.TryGetValue(c, out var rate))
                result[c] = rate;
        }
        return result;
    }

    public Dictionary<string, double> SearchCurrency(string query)
    {
        var allRates = GetRates();
        var result = new Dictionary<string, double>();
        var q = query.ToLower();
        foreach (var kv in allRates)
        {
            if (kv.Key.ToLower().Contains(q))
                result[kv.Key] = kv.Value;
        }
        return result;
    }

    public async Task<bool> Refresh() => await FetchRates();

    public bool ExportJSON(string filename = "rates.json")
    {
        try
        {
            var data = new
            {
                timestamp = lastUpdate?.ToString("o"),
                rates = GetRates()
            };
            var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(filename, json);
            return true;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Export failed: {e.Message}");
            return false;
        }
    }

    public bool ExportCSV(string filename = "rates.csv")
    {
        try
        {
            var rates = GetRates();
            var lines = new List<string> { "Currency,Rate" };
            foreach (var kv in rates.OrderBy(k => k.Key))
                lines.Add($"{kv.Key},{kv.Value:F4}");
            File.WriteAllLines(filename, lines);
            return true;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Export failed: {e.Message}");
            return false;
        }
    }

    public void ShowHistory()
    {
        if (history.Count == 0)
        {
            Console.WriteLine("No history yet.");
            return;
        }
        for (int i = 0; i < history.Count; i++)
        {
            Console.WriteLine($"[{i+1}] {history[i].Timestamp} – {history[i].Rates.Count} currencies");
        }
    }

    public void DisplayRates(Dictionary<string, double> rates = null)
    {
        if (rates == null) rates = GetRates();
        if (rates.Count == 0)
        {
            Console.WriteLine("No rates available.");
            return;
        }
        Console.WriteLine($"\nRates (USD base) – updated: {lastUpdate?.ToString("yyyy-MM-dd HH:mm:ss") ?? "unknown"}");
        Console.WriteLine(new string('-', 40));
        int count = 0;
        foreach (var kv in rates.OrderBy(k => k.Key))
        {
            if (count >= 20) break;
            Console.WriteLine($"{kv.Key,-5} : {kv.Value:F4}");
            count++;
        }
        if (rates.Count > 20)
            Console.WriteLine($"... and {rates.Count - 20} more");
    }

    static async Task Main()
    {
        var parser = new RateParser();
        Console.WriteLine("=== Exchange Rate Parser ===");
        while (true)
        {
            Console.WriteLine("\n1. Show all rates");
            Console.WriteLine("2. Filter by currency");
            Console.WriteLine("3. Search currency");
            Console.WriteLine("4. Export to JSON");
            Console.WriteLine("5. Export to CSV");
            Console.WriteLine("6. Refresh rates");
            Console.WriteLine("7. Show history");
            Console.WriteLine("8. Exit");
            Console.Write("Choose: ");
            var choice = Console.ReadLine()?.Trim() ?? "";
            switch (choice)
            {
                case "1":
                    parser.DisplayRates();
                    break;
                case "2":
                    Console.Write("Enter currency codes (comma-separated): ");
                    var input = Console.ReadLine()?.Trim().ToUpper() ?? "";
                    var currencies = input.Split(',').Select(c => c.Trim()).Where(c => !string.IsNullOrEmpty(c)).ToList();
                    if (currencies.Any())
                    {
                        var filtered = parser.FilterRates(currencies);
                        parser.DisplayRates(filtered);
                    }
                    else
                    {
                        Console.WriteLine("No currencies specified.");
                    }
                    break;
                case "3":
                    Console.Write("Enter currency code or name: ");
                    var query = Console.ReadLine()?.Trim() ?? "";
                    if (!string.IsNullOrEmpty(query))
                    {
                        var found = parser.SearchCurrency(query);
                        parser.DisplayRates(found);
                    }
                    else
                    {
                        Console.WriteLine("Query cannot be empty.");
                    }
                    break;
                case "4":
                    Console.Write("Filename (default: rates.json): ");
                    var fname = Console.ReadLine()?.Trim() ?? "rates.json";
                    if (string.IsNullOrEmpty(fname)) fname = "rates.json";
                    if (parser.ExportJSON(fname))
                        Console.WriteLine($"Exported to {fname}");
                    break;
                case "5":
                    Console.Write("Filename (default: rates.csv): ");
                    fname = Console.ReadLine()?.Trim() ?? "rates.csv";
                    if (string.IsNullOrEmpty(fname)) fname = "rates.csv";
                    if (parser.ExportCSV(fname))
                        Console.WriteLine($"Exported to {fname}");
                    break;
                case "6":
                    if (await parser.Refresh())
                        Console.WriteLine("Rates refreshed.");
                    else
                        Console.WriteLine("Refresh failed.");
                    break;
                case "7":
                    parser.ShowHistory();
                    break;
                case "8":
                    Console.WriteLine("Goodbye!");
                    return;
                default:
                    Console.WriteLine("Invalid choice.");
                    break;
            }
        }
    }
}
